src/sdk/bucketspace_fs.js

```diff
@@ -485,13 +496,49 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
     // BUCKET TAGGING //
     ////////////////////
 
-    async put_bucket_tagging(params) {
+    /**
+     * @param {*} params 
+     * @param {nb.ObjectSDK} object_sdk 
+     */
+    async put_bucket_tagging(params, object_sdk) {
         try {
             const { name, tagging } = params;
             dbg.log0('BucketSpaceFS.put_bucket_tagging: Bucket name, tagging', name, tagging);
-            const bucket = await this.config_fs.get_bucket_by_name(name);
-            bucket.tag = tagging;
-            await this.config_fs.update_bucket_config_file(bucket);
+            const bucket_path = this.config_fs.get_bucket_path_by_name(name);
+
+            const bucket_lock_file = `${bucket_path}.lock`;
+            await native_fs_utils.lock_and_run(this.fs_context, bucket_lock_file, async () => {
+                const bucket = await this.config_fs.get_bucket_by_name(name);
+                const { ns } = await object_sdk.read_bucket_full_info(name);
+                const is_bucket_empty = await BucketSpaceFS.#_is_bucket_empty(name, null, ns, object_sdk);
+
+                const tagging_object = bucket.tag?.reduce((curr, tag) => Object.assign(curr, { [tag.key]: tag.value }), {});
```

Comment:
I suggest moving the tagging logic to a function to keep this function simple and clean